templates/configmap.yaml

```diff
@@ -1,35 +1,53 @@
-{{- if .Values.slave.enabled -}}
 ---
 apiVersion: v1
 kind: ConfigMap
 metadata:
-  name: netdata-conf-slave
+  name: netdata-conf-master
   labels:
     app: {{ template "netdata.name" . }}
     chart: {{ template "netdata.chart" . }}
     release: {{ .Release.Name }}
     heritage: {{ .Release.Service }}
 data:
-  {{- range $name, $config := .Values.slave.configs }}
+  {{- range $name, $config := .Values.master.configs }}
   {{- if $config.enabled }}
   {{ $name }}: {{ toYaml $config.data | indent 4 }}
   {{- end }}
   {{- end }}
-{{- end }}
 
 ---
+
+{{- if .Values.slave.enabled -}}
 apiVersion: v1
 kind: ConfigMap
 metadata:
-  name: netdata-conf-master
+  name: netdata-conf-slave
   labels:
     app: {{ template "netdata.name" . }}
     chart: {{ template "netdata.chart" . }}
     release: {{ .Release.Name }}
     heritage: {{ .Release.Service }}
 data:
-  {{- range $name, $config := .Values.master.configs }}
+  {{- range $name, $config := .Values.slave.configs }}
   {{- if $config.enabled }}
   {{ $name }}: {{ toYaml $config.data | indent 4 }}
   {{- end }}
   {{- end }}
+{{- end }}
+
+---
+
+{{- if and (.Values.slave.enabled) (.Values.sd.slave.enabled) -}}
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: netdata-slave-sd-config-map
+  labels:
+    app: {{ template "netdata.name" . }}
+    chart: {{ template "netdata.chart" . }}
+    release: {{ .Release.Name }}
+    heritage: {{ .Release.Service }}
+data:
+  config.yml: |
+  {{- .Files.Get .Values.sd.slave.file | nindent 4 }}
```

Comment:
How can the chart user override this without modifying the chart?

Reply:
overwrite what?

file is configurable

https://github.com/ilyam8/helmchart/blob/b41183085a7c09fdde46bffefbc40074f2287f35/values.yaml#L14

Reply:
if you mean the file - copy paste the file, customize it to fit his need. It looks better then have all the configuration in the `values.yml` - there is only file name to use in the values.

Reply:
[i did it like that at first](https://github.com/netdata/helmchart/pull/102/commits/b19c21fc7e39cdbda6f2993e1eebea4a50761223) but it makes `values.yml` really big, having the config in a separate file serve same purpose and looks more clear to me.

Reply:
> if you mean the file - copy paste the file, customize it to fit his need

Common practice in cases like this is to allow the user to override the configmap name.

Reply:
can you elaborate how it works from your perspective?

i see we dont override config map names
> https://github.com/netdata/helmchart/blob/master/templates/configmap.yaml

configuration is a file, use can have its own configuration and override the  filename option. Why it doesnt do?

Reply:
like

```yaml
sd:
  repository: netdata/agent-sd
  tag: latest
  pullPolicy: Always
  slave:
    enabled: true
    file: sd-slave.yml
    configmap:
      name: <name>
      key: <name>
    resources: {}
```

And if `configmap.name` and `configmap.key` *are set* we dont generate our configmap and use custom user cmap?

Reply:
i was thinking about something like that, i decided not to do it because - it is easy to add options, it is hard to remove options, i decided to go with minimal config and if doesnt work for someone and there is a request to change/add something *only after it*, if reasoning is good - add it.

Reply:
> configuration is a file

User of helm shouldn't be expected to have a config file on their filesystem. For example, when using helm from terraform, it's almost never the case that a config file is first generated, saved to disk, and then referenced from a chart. More commonly, content of a file is passed to helm as value.

> a user can have its own configuration and override the filename option

I've never seen this done anywhere else and (not having actually tested it) don't even know if it works when installing a chart from a repository. Have you tried it?

> And if configmap.name and configmap.key are set we dont generate our configmap and use custom user cmap?

Yes exactly! But without the `file` option, as it no longer something the user cares about. As mentioned earlier, helm shouldn't rely on having access to filesystem of the user for values.

Reply:
> don't even know if it works when installing a chart

From [Helm Docs](https://helm.sh/docs/chart_template_guide/accessing_files/#lines):

> There is no way to pass files external to the chart during `helm install`

Reply:
All the templates are on the filesystem, there is just another file.

> Have you tried it?

There is documentation

>  But sometimes it is desirable to import a file that is not a template and inject its contents without sending the contents through the template renderer.

> https://helm.sh/docs/chart_template_guide/accessing_files/

You can check it works (test i did in [the first post](https://github.com/netdata/helmchart/pull/102#issue-427930112))

> helm template -s templates/configmap.yaml .

---

Let me summarize, you suggestion is:
 - put default configuration to the `values.yml`
 - generate config map using this config, the name should be hardcoded
 - have an option to set configmap name/key
 - if both are set do not generate configmap with the config from values and  just use these parametes ([here](https://github.com/ilyam8/helmchart/blob/b41183085a7c09fdde46bffefbc40074f2287f35/templates/daemonset.yaml#L176))

?

Reply:
>  installing a chart from a repository. Have you tried it?

i think i misunderstood you, what i did is `git clone/helm install`

Reply:
ill think how to change the PR

Reply:
will something like this do?

```yaml
sd:
  repository: netdata/agent-sd
  tag: latest
  pullPolicy: Always
  slave:
    enabled: true
    config: |
      ... data
    resources: {}
```

So a user can override just the whole config? Do we need cmap name/key configurable? Or you mean there is a chance the user might want to have this config in another configmap? not dedicated for the service?

Reply:
it would be very nice you show the config example you have in mind if you have time

Reply:
We provide our chart from repository as well. We can't expect users to clone the helm chart, modify it (ie add a new file in the chart directory) and *then* install it, in order to override config.

> Let me summarize, your suggestion is:

Almost. Here's the slightly modified one:

* Keep the `sd-slave.yml` as is (to reap the benefits of clean `values.yaml`)
* Reference it in the template as you do currently.
* Remove the `sd.slave.file` from `values.yaml`
* Generate config map using this config, the name should be hardcoded
* Have an option to set configmap name/key
* If both are set do not generate configmap with the config from values and just use these parameters

Reply:
> it would be very nice you show the config example you have in mind if you have time

You had it down perfectly before:

```yaml
sd:
  repository: netdata/agent-sd
  tag: latest
  pullPolicy: Always
  slave:
    enabled: true
    configmap:
      name: <name>
      key: <name>
    resources: {}
```

Reply:
i see you dont like `file` option, however not having it means i hardcode `sd-slave.yml` and there is no way to overwrite it (i think to configure a template is no-go), but if a had my own config and i use helm i would like to overwrite it, what if

```
sd:
  repository: netdata/agent-sd
  tag: latest
  pullPolicy: Always
  slave:
    enabled: true
    configmap:
      name: netdata-slave-sd-config-map
      key: config.yml
      file: sd-slave.yml
    resources: {}
```

if there is `file` we generate the map from file, if there is no file (or the value is `""`) we dont generate the map (in both cases we use name/key) ?

Reply:
it could be even:

- generate from file
- or generate from a value
- do not generate and use just name/key


```yml
sd:
  repository: netdata/agent-sd
  tag: latest
  pullPolicy: Always
  slave:
    enabled: true
    configmap:
      name: netdata-slave-sd-config-map
      key: config.yml
      from:
        file: sd-slave.yml
        value: ""
    resources: {}
```

Reply:
generating from a file (not hardcoded filename) needed at least by me

Reply:
There could be different application identification patterns, we could have several configs in the repo and a user could just switch between them.

Datadog uses container name for app identification for example

> https://docs.datadoghq.com/agent/kubernetes/integrations/?tab=kubernetes#configuration

Reply:
sure, that sounds reasonable

Reply:
thx for the help, i will do like that, i think it make sense to create a dir and put defaut config in there (better then have in the root)

Reply:
@ncmans i think i addressed your suggestions